drivers/SmartThings/sonos/src/init.lua

```diff
@@ -73,21 +83,94 @@ local function find_player_for_device(driver, device, should_continue)
   return player_found
 end
 
+local function update_fields_from_ssdp_scan(driver, device, fields)
+  local current_player_id = device:get_field(PlayerFields.PLAYER_ID)
+  local current_url = device:get_field(PlayerFields.WSS_URL)
+  local current_household_id = device:get_field(PlayerFields.HOUSEHOULD_ID)
+  local sonos_conn = device:get_field(PlayerFields.CONNECTION) --- @type SonosConnection
+
+  local already_connected = sonos_conn ~= nil
+  local refresh = false
+
+  if current_player_id ~= fields.player_id then
+    if current_player_id ~= nil then
+      driver._player_id_to_device[current_player_id] = nil
+      driver.sonos:mark_player_as_removed(current_player_id)
+      refresh = true
+    end
+    driver._player_id_to_device[fields.player_id] = device -- quickly look up device from player id string
+    driver.sonos:mark_player_as_joined(fields.player_id)
+    device:set_field(PlayerFields.PLAYER_ID, fields.player_id, { persist = true })
+  end
+
+  if current_household_id ~= fields.household_id then
+    if current_household_id ~= nil then
+      refresh = true
+    end
+    device:set_field(PlayerFields.HOUSEHOULD_ID, fields.household_id, { persist = true })
+  end
+
+  if current_url ~= fields.wss_url then
+    if current_url ~= nil and sonos_conn ~= nil then
+      sonos_conn:stop()
+      device:set_field(PlayerFields.CONNECTION, nil)
+      refresh = true
+    end
+    device:set_field(PlayerFields.WSS_URL, fields.wss_url, { persist = true })
+  end
+
+  if refresh and already_connected then
```

Comment:
is the expectation that the existing `sonos_conn` has already been stopped?

Reply:
Yeah, from #L113-120 above.

The general approach I was going for is:
1. Don't mess with the connection unless the websocket URL has changed
2. If the URL has changed (which is true if the existing URL is either nil or has a different value than the one in the scan), and if there's an existing sonos connection, then we stop it and nil it out.

However, typing it out, I see the *potential* bug. If we need to do a refresh but the connection wasn't changed, then we'll leak a new connection without stopping it.

The reason I missed this is because player_id and household_id are *supposed* to be stable/unchanging so in theory you should never be in a situation where you need to do a refresh unless the websocket URL changes. But I wanted to code very defensively here, and I think in that spirit we should probably also be defensive about that. Probably a `nil` check on the connection is sufficient.